read.js

```diff
@@ -0,0 +1,216 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+/* eslint max-statements:[0, 99] */
+'use strict';
+
+var TYPE = require('./TYPE');
+var ReadResult = require('bufrw/base').ReadResult;
+var bufrwErrors = require('bufrw/errors');
+var bufrw = require('bufrw');
+var errors = require('./errors');
+var TMapHeaderRW = require('./tmap').TMapRW.prototype.headerRW;
+var TListHeaderRW = require('./tlist').TListRW.prototype.headerRW;
+var StringRW = require('./string').StringRW;
+var BooleanRW = require('./boolean').BooleanRW;
+
+var widths = Object.create(null);
+widths[TYPE.VOID] = 0;
+
+var readVar = Object.create(null);
+readVar[TYPE.BOOL] = readBool;
+readVar[TYPE.BYTE] = readI8;
+readVar[TYPE.DOUBLE] = readDouble;
+readVar[TYPE.I8] = readI8;
+readVar[TYPE.I16] = readI16;
+readVar[TYPE.I32] = readI32;
+readVar[TYPE.I64] = readI64;
+readVar[TYPE.STRING] = readString;
+readVar[TYPE.STRUCT] = readStruct;
+readVar[TYPE.MAP] = readMap;
+readVar[TYPE.SET] = readList;
+readVar[TYPE.LIST] = readList;
+
+function readType(buffer, offset, typeid) {
+    var result;
+
+    // istanbul ignore else
+    if (readVar[typeid] !== undefined) {
+        result = readVar[typeid](buffer, offset);
+        // istanbul ignore if
+        if (result.err) {
+            return result;
+        }
+        offset = result.offset;
+
+    } else if (widths[typeid] !== undefined) {
+        var length = widths[typeid];
+        if (offset + length > buffer.length) {
+            return new ReadResult(bufrwErrors.ShortBuffer({
+                expected: offset + length,
+                actual: buffer.length,
+                buffer: buffer,
+                offset: offset
+            }));
+        }
+        offset += length;
+
+    } else {
+        return new ReadResult(errors.InvalidTypeidError({
+            typeid: typeid,
+            what: 'field::type'
+        }));
+    }
+
+    return new ReadResult(null, offset, result && result.value);
+}
+
+function readBool(buffer, offset) {
+    return BooleanRW.readFrom(buffer, offset);
+}
+
+function readDouble(buffer, offset) {
+    return bufrw.DoubleBE.readFrom(buffer, offset);
+}
+
+function readI8(buffer, offset) {
+    return bufrw.Int8.readFrom(buffer, offset);
+}
+
+function readI16(buffer, offset) {
+    return bufrw.Int16BE.readFrom(buffer, offset);
+}
+
+function readI32(buffer, offset) {
+    return bufrw.Int32BE.readFrom(buffer, offset);
+}
+
+function readI64(buffer, offset) {
+    return bufrw.FixedWidth(8).readFrom(buffer, offset);
+}
+
+function readString(buffer, offset) {
+    return StringRW.readFrom(buffer, offset);
+}
```

Comment:
readString = StringRW.readFrom. No need for wrapping for cases like these. readFrom does not use StringRW as a holder of state.